src/Bash/Simplify.hs

```diff
@@ -74,18 +74,50 @@ replace' somethingElse = return somethingElse
 
 replaceStr :: BashString -> State Namespace BashString
 replaceStr s@(SingleQ _) = return s
-replaceStr (DoubleQ s)   = DoubleQ `fmap` replaceStr' s
-replaceStr (NoQuote s)   = NoQuote `fmap` replaceStr' s
+replaceStr (DoubleQ s)   = expandStr s >>= \strs -> return $ DoubleQ $ map Right strs
+replaceStr (NoQuote s)   = expandStr s >>= \strs -> return $ NoQuote $ map Right strs
 replaceStr (Backtic f)   = Backtic `fmap` replace' f
 
--- | Doesn't yet support ${foo[...]}
-replaceStr' :: String -> State Namespace String
-replaceStr' s = get >>= \ns ->
-   case s =~ "\\${?[\\w]+}?" :: (String,String,String) of
-     (_,"",_) -> return s
-     (b,m,a)  -> ((b ++ replaced) ++) `fmap` replaceStr' a
-         where replaced = fromMaybe m (head `fmap` getVar ns m')
-               m'       = filter (`notElem` "${}") m
+expandStr :: [Either BashExpansion String] -> State Namespace [String]
+expandStr [] = return []
+expandStr (x:xs) = do
+  lhs <- expandStr' x
+  rhs <- expandStr xs
+  return $ lhs ++ rhs
+
+expandStr' :: Either BashExpansion String -> State Namespace [String]
+expandStr' (Right s) = return [s]
+expandStr' (Left (BashExpansion var sub)) = do
+  subscript <- mapM expandWrapper sub
+  ns <- get
+  getIndexedString ns var $ foldr (++) "" subscript
+
+getIndexedString :: Namespace -> String -> String -> State Namespace [String]
+getIndexedString ns var idx  = case  M.lookup var ns  of
+  Nothing -> return []
+  Just bstrs -> getIndexedString' bstrs idx
+
+getIndexedString' :: [BashString] -> String -> State Namespace [String]
+getIndexedString' bstrs "" = do
+  singlestr <- replaceStr $ head bstrs
+  return $ fromBashString singlestr
+getIndexedString' bstrs "*" = do
+  strings <- mapM replaceStr bstrs
+  return $ foldr (++) [] $ map fromBashString strings
+getIndexedString' bstrs "@" = getIndexedString' bstrs "*"
+getIndexedString' bstrs num = do
+  singlestr <- replaceStr $ bstrs !! read num
+  return $ fromBashString singlestr
+
+expandWrapper :: BashString -> State Namespace String
+expandWrapper (SingleQ s) = return s
+expandWrapper (DoubleQ s) = do
+  strs <- expandStr s
+  return $ foldr (++) "" strs
+expandWrapper (NoQuote s) = do
+  strs <- expandStr s
+  return $ foldr (++) "" strs
+expandWrapper (Backtic f) = replace' f >>= \x -> return $ unwords $ fromCommand x
```

Comment:
I meant for the project as a whole. I thought you were thinking of going through *everything*. Sorry. :) A separate PR would work better there because it's a separate discussion. I'm not quite sure what general aura policy on when you switch to do-notation is, and I'm currently on mobile and not able to check. I think it would be worth a discussion, however. What you may want to do is just use do-notation for now and then PR the style change with this as a sample. As for indentation problems, fixes are always good.